### ISSUE TITLE
What can we implement from Karpathy's "Best ChatGPT $100 can buy"

### ISSUE THREAD

**Issue by vukrosic** [Open-Superintelligence-Lab/5-dollar-llm#25]
Check this repo and suggest what we can implement - https://github.com/karpathy/nanochat

**Comment by shehab-ashraf** [comment]
@vukrosic I propose switching to a fan-based initialization scheme that adapts to layer dimensions.
```
def _init_weights(self, module):
    if isinstance(module, nn.Linear):
        # Adaptive initialization based on layer dimensions
        # Reference: https://arxiv.org/pdf/2310.17813
        fan_out = module.weight.size(0)
        fan_in = module.weight.size(1)
        std = 1.0 / math.sqrt(fan_in) * min(1.0, math.sqrt(fan_out / fan_in))
        torch.nn.init.normal_(module.weight, mean=0.0, std=std)
        if module.bias is not None:
            torch.nn.init.zeros_(module.bias)
    elif isinstance(module, nn.Embedding):
        torch.nn.init.normal_(module.weight, mean=0.0, std=1.0)
```

**Comment by shehab-ashraf** [comment]
> is this all it is? if yes I can copy paste it into my AI, i need to benchmark and if we keep it I will credit you

ok here's changes we should do it in intialization:

Our current initialization uses  (fixed std=0.02, no zero-init for residuals). This will cause slower convergence and not best for training stability.

### 1. Fixed std=0.02:

Current code:
```
torch.nn.init.normal_(module.weight, mean=0.0, std=0.02
```
**problem**:
- Ignores layer dimensions (fan-in/fan-out)
- GQA layers (k_proj, v_proj) have different dimensions than q_proj - same std is wrong
- Causes activation magnitude drift across layers
**Solution**: 
- Use muP-inspired fan-based initialization
```
def _init_weights(self, module):
    if isinstance(module, nn.Linear):
        fan_in = module.weight.size(1)
        fan_out = module.weight.size(0)
        std = 1.0 / math.sqrt(fan_in) * min(1.0, math.sqrt(fan_out / fan_in))
        torch.nn.init.normal_(module.weight, mean=0.0, std=std)
        if module.bias is not None:
            torch.nn.init.zeros_(module.bias)
    elif isinstance(module, nn.Embedding):
        torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
```

### 2. No Zero-Init for Residual Paths:

**Current behavio**r:
Each transformer block adds random noise at initialization: x = x + random_transformation(x)

**Problems**:
- Training instability and slow convergence.

**Solution**:
Zero-initialize all residual output projections.
```
def _init_special_weights(self):
    """Zero-init residual projections for training stability"""
    # Zero out the output head
    torch.nn.init.zeros_(self.lm_head.weight)
    
    # Zero out residual projections in each transformer block
    for block in self.transformer_blocks:
        torch.nn.init.zeros_(block.attention.w_o.weight)
        torch.nn.init.zeros_(block.feed_forward.down_proj.weight)
```
**With zero-init**:
- Input: embedding vectors from self.token_embedding(x)
Output after blocks: the same embedding vectors (unchanged)
- the network starts simple (identity) and learns complexity gradually, rather than starting with random.

### 3. Embedding Scaling(why?):
Conflicts with RMSNorm (we normalize immediately after)
Modern models (GPT-3, Llama, Mistral) don't use this
i think we should remove the scaling: 
```
x = self.token_embedding(x)  # Remove * math.sqrt(self.config.d_model)
```

### 4. (Optional) Memory Optimization

Cast embeddings to BF16 to save memory
```
def _init_special_weights(self):
    # ... (zero-init code from above) ...
    
    # Cast embeddings to BF16
    if self.token_embedding.weight.device.type == "cuda":
        self.token_embedding.to(dtype=torch.bfloat16)
```
memory saved with minimal quality loss
